# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1792920859 -3600
# Parent  c8d89da3e41818772283ef770f3b37935d96651d
Bug 2002774 - Fix lagging status bar.

In 140 this was a backout bug 1901846 for generally being the wrong solution and causing bug 1945489.
Bug 1945489 got side-tracked and the remaining issue is now in bug 2002774.

The patch will now only fix showStatusString() in mailWindow.js by reverting it to the fine
implementation from bug 1864776
https://hg-edge.mozilla.org/comm-central/rev/2d7e8092cbc4
before bug 1945489 got it all wrong.

The original code is our code from
https://github.com/Betterbird/thunderbird-patches/blob/main/115/bugs/1864776-status-display-less-busy.patch
now in
https://github.com/Betterbird/thunderbird-patches/blob/main/115/unused/1864776-status-display-less-busy.patch

In 2026 we increased the time for clearing and removed a needless reset.

diff --git a/mail/base/content/mailWindow.js b/mail/base/content/mailWindow.js
--- a/mail/base/content/mailWindow.js
+++ b/mail/base/content/mailWindow.js
@@ -390,16 +390,17 @@ nsMsgStatusFeedback.prototype = {
 
   // Member variables.
   _startTimeoutID: null,
   _stopTimeoutID: null,
+  _clearTimeoutID: null,
   // How many start meteors have been requested.
   _startRequests: 0,
   _meteorsSpinning: false,
   _progressBarVisible: false,
   _activeProcesses: null,
   _statusFeedbackProgress: -1,
   _statusLastShown: 0,
-  _statusQueue: null,
+  _lastStatusText: null,
   _timeoutDelay: ChromeUtils.isInAutomation ? 50 : 500,
 
   // unload - call to remove links to listeners etc.
   unload() {
@@ -489,49 +490,46 @@ nsMsgStatusFeedback.prototype = {
    * @param {string} statusText - The status string to display.
    * @see {nsIMsgStatusFeedback}
    */
   showStatusString(statusText) {
+    if (this._clearTimeoutID) {
+      // Cancel clear regardless:
+      // If we get a new status, we don't want to clear it with a stale timer.
+      // If we get an empty status, clearing is done already.
+      clearTimeout(this._clearTimeoutID);
+      this._clearTimeoutID = null;
+    }
+
     // Let's make sure the display doesn't flicker.
-    const MIN_DISPLAY_TIME = 750;
-    if (
-      !this._statusIntervalId &&
-      Date.now() - this._statusLastShown > MIN_DISPLAY_TIME
-    ) {
-      this._statusLastShown = Date.now();
+    const timeBetweenDisplay = 500;
+    const now = Date.now();
+    if (now - this._statusLastShown > timeBetweenDisplay) {
+      // Cancel any pending status message. The timeout is not guaranteed
+      // to run within timeBetweenDisplay milliseconds.
+      this._lastStatusText = null;
+
+      this._statusLastShown = now;
       if (this._statusText.value != statusText) {
         this._statusText.value = statusText;
       }
-      return;
-    }
-    if (this._statusQueue.length >= 10) {
-      // Drop further messages until the queue has cleared up.
-      return;
-    }
-    if (!statusText && this._statusQueue.length) {
-      // Don't queue empty strings in the middle.
-      return;
-    }
-    this._statusQueue.push(statusText);
-    if (this._statusIntervalId) {
-      return;
-    }
-
-    // Arrange for this to be shown in MIN_DISPLAY_TIME ms.
-    this._statusIntervalId = setInterval(() => {
-      const text = this._statusQueue.shift();
-      if (text === undefined) {
-        clearInterval(this._statusIntervalId);
-        this._statusIntervalId = null;
-        if (!this._meteorsSpinning) {
-          this._statusText.value = ""; // Clear status text once done.
-        }
+    } else {
+      if (this._lastStatusText !== null) {
+        // There's already a pending display. Replace it.
+        this._lastStatusText = statusText;
         return;
       }
-      this._statusLastShown = Date.now();
-      if (this._statusText.value != text) {
-        this._statusText.value = text;
-      }
-    }, MIN_DISPLAY_TIME);
+      // Arrange for this to be shown in timeBetweenDisplay milliseconds.
+      this._lastStatusText = statusText;
+      setTimeout(() => {
+        if (this._lastStatusText !== null) {
+          this._statusLastShown = Date.now();
+          if (this._statusText.value != this._lastStatusText) {
+            this._statusText.value = this._lastStatusText;
+          }
+          this._lastStatusText = null;
+        }
+      }, timeBetweenDisplay);
+    }
   },
 
   _startMeteors() {
     this._meteorsSpinning = true;
@@ -574,9 +572,10 @@ nsMsgStatusFeedback.prototype = {
     }
   },
 
   _stopMeteors() {
-    this.showStatusString("");
+    // Don't mix status string with meteors. It will be cleared indepenently.
+    // this.showStatusString("");
 
     // stop the throbber
     if (this._throbber) {
       this._throbber.setAttribute("busy", false);
@@ -594,18 +593,28 @@ nsMsgStatusFeedback.prototype = {
     // Update the stop button
     goUpdateCommand("cmd_stop");
   },
 
-  stopMeteors() {
+  stopMeteors(clearDelay = 2000) {
     if (this._startRequests > 0) {
       this._startRequests--;
     }
 
     // If we are going to be starting the meteors, cancel the start.
     if (this._startRequests == 0 && this._startTimeoutID) {
       clearTimeout(this._startTimeoutID);
       this._startTimeoutID = null;
-      this.showStatusString("");
+    }
+
+    // stopMeteors() will also clear the status message.
+    if (this._startRequests == 0 && !this._clearTimeoutID) {
+      this._clearTimeoutID = setTimeout(
+        () => {
+          this._clearTimeoutID = null;
+          this.showStatusString("");
+        },
+        clearDelay // Leave 2 seconds to read status message.
+      );
     }
 
     // If we have no more pending starts and we don't have a stop timeout
     // already in progress AND the meteors are currently running then fire a
@@ -1242,10 +1251,14 @@ contentProgress.addListener({
           window.MsgStatusFeedback.showStatusString(
             messengerBundle.GetStringFromName("documentLoading")
           );
         } else if (stateFlags & Ci.nsIWebProgressListener.STATE_STOP) {
-          window.MsgStatusFeedback.stopMeteors();
-          window.MsgStatusFeedback.showStatusString("");
+          // stopMeteors() already resets the status string, but with a delay.
+          // We use a shorter delay, so we don't see "Loading message..." for
+          // longer than necessary. Clearing the status string immediately
+          // might remove the previous status feedback too soon.
+          window.MsgStatusFeedback.stopMeteors(500);
+          // window.MsgStatusFeedback.showStatusString("");
         }
       }
     }
   },
